### PR TITLE
Add new SetMDFrames algorithm

### LINKED ISSUE #14239: Create SetMDFrames algorithm
In order to correct legacy files without knowledge of the MDFrames system, we require a fix up algorithm

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Fixes #14239 

OSX and Windows installers from the last Jenkins build can be found here: \olympic\Babylon5\Scratch\Anton\Testing\14239_SetMDFrames
# For testing
1. Code review
2. Check that the documentation in the rst file is understandable.
3. Open the SetMDFrames dialog and check that it looks alright
### Test instructions

The tests are based on the following script

``` python
# Prepare the sample workspace
inputFrame = 'QSample,QSample,QSample'
inputUnits = 'U,U,U'
outputFrame = 'QLab'
mdws = CreateMDWorkspace(Dimensions=3, Extents='-10,10,-10,10,-10,10', 
                                        Names='A,B,C', Units=inputUnits, 
                                        Frames=inputFrame)
print "The first MDFrame is of type " + mdws.getDimension(0).getMDFrame().name()
print "The second MDFrame is of type " + mdws.getDimension(0).getMDFrame().name()
print "The third MDFrame is of type " + mdws.getDimension(0).getMDFrame().name()

 # Set the MDFrames to new values
SetMDFrame(mdws, MDFrame=outputFrame, Axes=[0,2])
print
print "The new first MDFrame is of type " + mdws.getDimension(0).getMDFrame().name()
print "The second MDFrame is of type " + mdws.getDimension(1).getMDFrame().name() 
print "The new second MDFrame is of type " + mdws.getDimension(2).getMDFrame().name()
```

These tests are similar to the unit tests, but might give a starting point for random testing

**Test regular input and output_**
1. Set the input Frame to 'QSample,QSample,QSample'
2. Set the input Units to 'U,U,U'
3. Set the output frame to 'QLab'
4. Run the script
   - Confirm that QLab frame was chosen (should be printed out) for index 0 and index 2

**Test HKL selection with bad units**
1. Set the input Frame to 'General Frame,General Frame,General Frame'
2. Set the input Units to 'isdfsdf, gg, sdfdf'
3. Set the output frame to "HKL"
4. Run the script
   - Confirm that an error is generated

**Test HKL selection with bad units**
1. Set the input Frame to  'General Frame,General Frame,General Frame'
2. Set the input Units to 'in 2.6437 A^-1,in 2.6437 A^-1,in 2.6437 A^-1'
3. Set the output frame to "HKL"
4. Run the script
   - Confirm that the output frame is HKL (should be printed out) for index 0 and index 2

**Test incorrect selection**
1. Set the input Frame to 'General Frame,General Frame,General Frame'
2. Set the input Units to  'in 2.6437 A^-1,in 2.6437 A^-1,in 2.6437 A^-1'
3. Set the output frame to "sdfsdf"
4. Run the script
   - Confirm that an error is generated

### CODE REVIEW

File: Framework/MDAlgorithms/src/SetMDFrames.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+/**
+ * Creates an MDFrame based on the users selection
+ * @param frameSelection :: the selected frame type
+ * @param oldFrame :: a refrence to the frame we want to replace
+ * @returns :: a unique pointer to the new MDFrame
+ */
+Mantid::Geometry::MDFrame_uptr
+SetMDFrames::createMDFrame(const std::string &frameSelection,
```
Comment (OwenArnold) [inline Framework/MDAlgorithms/src/SetMDFrames.cpp]: Not sure why this needs to be a member function

File: Framework/MDAlgorithms/src/SetMDFrames.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+/**
+ * Check the inputs for invalid values
+ * @returns A map with validation warnings.
+ */
+std::map<std::string, std::string> SetMDFrames::validateInputs() {
+  std::map<std::string, std::string> invalidProperties;
+  Mantid::API::IMDWorkspace_sptr ws = getProperty("InputWorkspace");
+
+  if (!boost::dynamic_pointer_cast<Mantid::API::IMDEventWorkspace>(ws) &&
```
Comment (OwenArnold) [inline Framework/MDAlgorithms/src/SetMDFrames.cpp]: Fine

File: Framework/Geometry/inc/MantidGeometry/MDGeometry/MDHistoDimension.h [hunk 3]
```diff
@@ -113,6 +113,12 @@ class MANTID_GEOMETRY_DLL MDHistoDimension : public IMDDimension {
     m_binWidth = (m_max - m_min) / static_cast<coord_t>(m_numBins);
   }
 
+  /**
+   * Set the MDFrame
```
Comment (OwenArnold) [inline Framework/Geometry/inc/MantidGeometry/MDGeometry/MDHistoDimension.h]: I think we should be adding some text to this later to warn against using this. Setting the frame on an existing MDHistoDimension should not be a common use-case.

File: Framework/MDAlgorithms/src/SetMDFrames.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+void SetMDFrames::init() {
+  declareProperty(new WorkspaceProperty<Mantid::API::IMDWorkspace>(
+                      "InputWorkspace", "", Direction::InOut),
+                  "The workspace for which the MDFrames are to be changed. "
+                  "Note that only MDHisto and MDEvent workspaces can be "
+                  "altered by this algorithm.");
+
+  // Options for the MDFrames
+  std::vector<std::string> mdFrames;
+  mdFrames.push_back(Mantid::Geometry::GeneralFrame::GeneralFrameName);
```
Comment (OwenArnold) [inline Framework/MDAlgorithms/src/SetMDFrames.cpp]: This is neat

**Comment by OwenArnold** [comment]
This all looks OK, well tested and documented. 

However I have two questions/considerations.

1. I quite like the numpy style syntax we've been using recently. This would make the above something like. 
```
SetMDFrames(InputWorkspace=mdws,MDFrame='HKL', Indexes=[0,1,2])
```
I raise this now, because if the above is preferred, we need to implement it before we make the algorithm public. 
1. Given the above do you thinks we should call the function SetMDFrame rather than making it plural

**Comment by picatess** [comment]
* Changed naming to MDFrame
* Implemented the Numpy-styly syntax
* Added usage warning to setMDFrame on MDHistDimension

**Comment by OwenArnold** [comment]
@picatess I've carefully been through this looks solid. I'll merge this when we get green builds.

Might I make some suggestions for future enhancements. These would be new tickets if agreed.
 1. The verification looks very much like what we have in *TransposeMD* do you thinks we should share some of the implementation. Maybe via a base algorithm? This would make a good task for one of the new starters.
 1. So that users don't have to remember to use *SetSpecialCoordinates* for `IPeaksWorkspaces`, and now *SetMDFrames* for `MD*Workspaces`, do you think we should now deprecate the former and move the functionality into SetMDFrames?

**Comment by picatess** [comment]
@OwenArnold 
1. I agree some of the verification could be shared between these algorihtms, potentially even more. We would have to check. Having a base class would in a way indicate that `TransposeMD` and `SetMDFrame` are related though. We could outsource the checking to an new class and introduce it to   `TransposeMD` and `SetMDFrame`  via composition instread of inheritance. I think this is a good issue for the new starters. I will create a ticket for it once SetMDFrame is in master.
2. I think that is actually a good idea, since they are doing similar things albeit with different mechanisms. The issue I see is that we would be mixing `SpecialCoordinateSystem` which makes sense for Peaks-Workspaces but not for MDEvent- or MDHisto-Workspaces with `MDFrame` which makes sense for MDEvent- or MDHisto-Workspace but not for Peaks-Workspaces (at least as far as I understood it). That might lead to some confusion, but should be managable with sufficiently generic naming.

**Comment by OwenArnold** [comment]
@picatess 
 1. This would be a base class without any virtual behaviour, but I agree that the composition route is probably better.

**Pull request merged**